api/server/router/system/system_routes.go

```diff
@@ -90,49 +101,80 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
 }
 
 func (s *systemRouter) getDiskUsage(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
-	eg, ctx := errgroup.WithContext(ctx)
-
-	var du *types.DiskUsage
-	eg.Go(func() error {
-		var err error
-		du, err = s.backend.SystemDiskUsage(ctx)
+	if err := httputils.ParseForm(r); err != nil {
 		return err
-	})
+	}
+
+	var getContainers, getImages, getVolumes, getBuildCache bool
+	typesStrs, ok := r.Form["types"]
+	if !ok {
+		getContainers, getImages, getVolumes, getBuildCache = true, true, true, true
+	} else {
+		for _, s := range typesStrs {
+			for _, typ := range strings.Split(s, ",") {
+				switch typ {
+				case "container", "containers":
+					getContainers = true
+				case "image", "images":
+					getImages = true
+				case "volume", "volumes":
+					getVolumes = true
+				case "build", "build-cache", "build-caches":
+					getBuildCache = true
+				default:
+					return invalidRequestError{fmt.Errorf("unknown object type: %s", typ)}
+				}
+			}
+		}
+	}
 
-	var buildCache []*types.BuildCache
-	eg.Go(func() error {
+	var du *types.DiskUsage
+	diskUsage := func() error {
 		var err error
-		buildCache, err = s.builder.DiskUsage(ctx)
+		du, err = s.backend.SystemDiskUsage(ctx, DiskUsageOptions{
+			Containers: getContainers,
+			Images:     getImages,
+			Volumes:    getVolumes,
+		})
 		if err != nil {
-			return pkgerrors.Wrap(err, "error getting build cache usage")
+			return pkgerrors.Wrap(err, "error getting system disk usage")
 		}
 		return nil
-	})
-
-	if err := eg.Wait(); err != nil {
-		return err
 	}
 
-	var builderSize int64
-	for _, b := range buildCache {
-		builderSize += b.Size
-	}
+	if !getBuildCache {
+		if err := diskUsage(); err != nil {
+			return err
+		}
+	} else {
+		eg, ctx := errgroup.WithContext(ctx)
 
-	du.BuilderSize = builderSize
-	du.BuildCache = buildCache
+		eg.Go(diskUsage)
 
-	return httputils.WriteJSON(w, http.StatusOK, du)
-}
+		var buildCache []*types.BuildCache
+		eg.Go(func() error {
+			var err error
+			buildCache, err = s.builder.DiskUsage(ctx)
+			if err != nil {
+				return pkgerrors.Wrap(err, "error getting build cache usage")
```

Comment:
If we _do_ want to add the wrapping above, this one should also be included / have the same prefix;

```suggestion
				return pkgerrors.Wrap(err, "error getting disk usage: error getting build cache usage")
```

I'm wondering why this code is separate here, and why it was not included in `backend.SystemDiskUsage`; perhaps because they didn't want the builder to be added to the daemon struct 🤔 It would make things somewhat cleaner here though.

Reply:
> I'm wondering why this code is separate here, and why it was not included in `backend.SystemDiskUsage`; perhaps because they didn't want the builder to be added to the daemon struct thinking It would make things somewhat cleaner here though.

I'd guess it has something to do with https://github.com/moby/moby/blob/5b708507ff45019673d0ac92797edcdcc2908f00/daemon/disk_usage.go#L15-L18